### PR TITLE
feat(v4-sdk): Add V4 PositionManager SDK skeleton

### PR DESCRIPTION
**Pull request by zhongeric**
## Description

Adds a skeleton implementation of the Position Manager for V4. It also adds the Position entity, which is basically copied over from V3 but with some type changes to allow for the use of native currencies. For the full implementation, see #86 

And enumerating through the cases we want to cover:
Create pool -> createCallParameters
Create pool, mint position (with liquidity) -> addCallParameters with flags
Mint position with existing pool -> addCallParameters with flags
Add liquidity to existing position in existing pool -> addCallParameters
Decrease liquidity from position and burn the position -> removeCallParameters with flags
Decrease liquidity from position -> removeCallParameters
Collect fees from position -> collectCallParameters

## How Has This Been Tested?

_[e.g. Manually, E2E tests, unit tests, Storybook]_

## Are there any breaking changes?

_[e.g. Type definitions, API definitions]_

If there are breaking changes, please ensure you bump the major version Bump the major version (by using the title `feat(breaking): ...`), post a notice in #eng-sdks, and explicitly notify all Uniswap Labs consumers of the SDK.

## (Optional) Feedback Focus

_[Specific parts of this PR you'd like feedback on, or that reviewers should pay closer attention to]_

## (Optional) Follow Ups

_[Things that weren't addressed in this PR, ways you plan to build on this work, or other ways this work could be extended]_

### CODE REVIEW

File: sdks/v4-sdk/src/posm.ts [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+  /**
+   * The account that should receive the tokens.
+   */
+  recipient: string
+}
+
+export type MintOptions = CommonOptions & CommonAddLiquidityOptions & MintSpecificOptions
+export type IncreaseOptions = CommonOptions & CommonAddLiquidityOptions & ModifyPositionSpecificOptions
+
+export type DecreaseLiquidityOptions = CommonOptions & RemoveLiquiditySpecificOptions & ModifyPositionSpecificOptions
```
Comment (snreynolds) [inline sdks/v4-sdk/src/posm.ts:131]: BurnLiquidityOptions?

File: sdks/v4-sdk/src/posm.ts [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+  }
+
+  /**
+   * Produces the calldata for completely or partially exiting a position
+   * @param position The position to exit
+   * @param options Additional information necessary for generating the calldata
+   * @returns The call parameters
+   */
+  public static removeCallParameters(_position: Position, _options: DecreaseLiquidityOptions): MethodParameters {
+    // Encode DECREASE_LIQUIDITY and then COLLECT
```
Comment (snreynolds) [inline sdks/v4-sdk/src/posm.ts:205]: this will be either a decrease or burn and then an encoding of take pair

File: sdks/v4-sdk/src/posm.ts [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+  expectedCurrencyOwed1: CurrencyAmount<Currency>
+
+  /**
+   * The account that should receive the tokens.
+   */
+  recipient: string
+}
+
+export type MintOptions = CommonOptions & CommonAddLiquidityOptions & MintSpecificOptions
+export type IncreaseOptions = CommonOptions & CommonAddLiquidityOptions & ModifyPositionSpecificOptions
```
Comment (snreynolds) [inline sdks/v4-sdk/src/posm.ts:129]: I think CommonAddLiquidityOptions could share the slippageTolerance param?

Is that percentage applied to each amount then? (amount0 and amount1) ?

File: sdks/v4-sdk/src/posm.ts (hunk 1 above)
Comment (snreynolds) [inline sdks/v4-sdk/src/posm.ts:131]: Where is the decrease/collect slippageTolerance param?

**Review by snreynolds — COMMENTED** [review]
are you pulling the utils from v3-sdk?

and then I think we can separately have a PR for encoding all the actions with a v4 base actions planner?

File: sdks/v4-sdk/src/posm.ts (hunk 3 above)
Comment (zhongeric) [reply sdks/v4-sdk/src/posm.ts:129]: yeah i think its applied to each amount in V3

File: sdks/v4-sdk/src/posm.ts (hunk 1 above)
Comment (zhongeric) [reply sdks/v4-sdk/src/posm.ts]: Ah added to CommonOptions

**Comment by zhongeric** [comment]
I think we should implement the actions router in #87 , which can be separate from this PR which doesn't do any implementation work

File: sdks/v4-sdk/src/posm.ts (hunk 1 above)
Comment (snreynolds) [reply sdks/v4-sdk/src/posm.ts]: Sorry I just meant are there going to be specific ones for burn? You could use RemoveLiquiditySpecificOptions I guess but there is no need for a liquidityPercentage to be passed in

File: sdks/v4-sdk/src/entities/position.ts [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+   * Returns the price of token0 at the upper tick
+   */
+  public get token0PriceUpper(): Price<Currency, Currency> {
+    throw new Error('Not implemented')
+  }
+
+  /**
+   * Returns the amount of token0 that this position's liquidity could be burned for at the current pool price
+   */
+  public get amount0(): CurrencyAmount<Currency> {
```
Comment (snreynolds) [inline sdks/v4-sdk/src/entities/position.ts:73]: doesn't this calculation need the current price and price at lower and price at upper?

Comment (snreynolds) [reply sdks/v4-sdk/src/entities/position.ts:73]: discussed offline - will do similar to v3 impl.

File: sdks/v4-sdk/src/posm.ts (hunk 1 above)
Comment (snreynolds) [reply sdks/v4-sdk/src/posm.ts]: under the hood though if a user wants to burn we should just encode a BURN call not a decrease THEN burn

File: sdks/v4-sdk/src/utils/abi.ts [hunk 5]
```diff
@@ -0,0 +1,524 @@
+// TODO: import this from npm
```
Comment (snreynolds) [inline sdks/v4-sdk/src/utils/abi.ts:1]: where would you publish this?

File: sdks/v4-sdk/src/utils/actions.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
+import { defaultAbiCoder } from 'ethers/lib/utils'
+
```
Comment (snreynolds) [inline sdks/v4-sdk/src/utils/actions.ts:2]: would be nice to keep router + posm actions using the same `encodeAction` code so whenever that gets merged maybe we re architect this so we have posm actions, router actions and then they share a "planner" file that does the encoding/building of actions

Comment (zhongeric) [reply sdks/v4-sdk/src/utils/actions.ts:2]: yea i agree, ok we can wait on #87 for this one then. Talked to Ayo and they're ok with that

**Review by snreynolds — APPROVED** [review]
nice!

File: sdks/v4-sdk/src/utils/abi.ts (hunk 5 above)
Comment (zhongeric) [reply sdks/v4-sdk/src/utils/abi.ts:1]: hmmm I'm not entirely sure but the v3-periphery abis are published in npm

File: sdks/v4-sdk/src/utils/actions.ts (hunk 6 above)
Comment (zhongeric) [reply sdks/v4-sdk/src/utils/actions.ts]: ok actually going to remove this to unblock them

File: sdks/v4-sdk/src/entities/position.ts [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    amount1,
+    useFullPrecision,
+  }: {
+    pool: Pool
+    tickLower: number
+    tickUpper: number
+    amount0: BigintIsh
+    amount1: BigintIsh
+    useFullPrecision: boolean
+  }) {
```
Comment (dannydaniil) [inline sdks/v4-sdk/src/entities/position.ts:136]: could we please add return types for this one and below?

**Review by dannydaniil — COMMENTED** [review]
looks great, just added 1 comment

**Review by snreynolds — APPROVED** [review]
looks good to me

**Review by dannydaniil — APPROVED** [review]
lgtm

**Pull request merged**